system_tests/nodeinterface_test.go

```diff
@@ -0,0 +1,72 @@
+// Copyright 2021-2022, Offchain Labs, Inc.
+// For license information, see https://github.com/nitro/blob/master/LICENSE
+
+package arbtest
+
+import (
+	"context"
+	"math/big"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/offchainlabs/nitro/arbos/util"
+	"github.com/offchainlabs/nitro/solgen/go/node_interfacegen"
+)
+
+func getL1BlockNum(t *testing.T, ctx context.Context, client *ethclient.Client, l2BlockNum uint64) uint64 {
+	header, err := client.HeaderByNumber(ctx, big.NewInt(int64(l2BlockNum)))
+	Require(t, err)
+	l1BlockNum := types.DeserializeHeaderExtraInformation(header).L1BlockNumber
+	return l1BlockNum
+}
+
+func TestGetL2BlockRangeForL1(t *testing.T) {
+	t.Parallel()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	l2info, node, l2client, l1info, _, _, l1stack := createTestNodeOnL1(t, ctx, true)
+	defer requireClose(t, l1stack)
+	defer node.StopAndWait()
+	user := l1info.GetDefaultTransactOpts("User", ctx)
+
+	numTransactions := 30
+	for i := 0; i < numTransactions; i++ {
+		TransferBalanceTo(t, "Owner", util.RemapL1Address(user.From), big.NewInt(1e18), l2info, l2client, ctx)
+	}
+
+	nodeInterface, err := node_interfacegen.NewNodeInterface(types.NodeInterfaceAddress, l2client)
+	Require(t, err)
+
+	l1BlockNums := map[uint64][]uint64{}
```

Comment:
You only need to store min/max (first/last) right? Why store all the block ids ?

Reply:
This is still slice of arbitrary size here instead of just first/last.

Reply:
that is because some L1 numbers can just appear in one L2 block and in that case the range would be `[x, x]` and in these kind of cases cases the length of list would be 1 so we would need to have a arbitrary length slice to handle this. And one thing to note here is the size of slices never exceed 2, so its equivalent to map[uint64][2]uint64{}

Reply:
That is correct, the size of slices in your approach never exceed 2, and my point is you are using unsuitable struct to store your data.

It's fully in control how you construct that map. It all boils down to a simple problem:
- you have a stream of pairs: [l1BlockNum, l2BlockNum] and need to construct a map M where M[l1BlockNum] indicates min and max l2BlockNum for it.
You can use [2]uint64{} (instead of implying that algorithm basically only needs that), or even introduce your struct to store high/low indices do it for example like this:

```lang=go
type pair struct {
	low, high uint64
}

// Maps l1 block number to a pair of min and max l2 block numbers that corresponding l1 block contains.
m := make(map[uint64]*pair)
for l2 := uint64(0); l2 <= latestL2; l2++ {
	l2 := l1BlockNum(l2)
        if _, exists := m[l1]; !exists {
	m[l1] = &pair{low:l2, high:l2}
        }
        m[l1].high = l2
}